### PR TITLE
Get new recipe format and planner into a state where we can iterate on them.

### PR DESCRIPTION
**Pull request by shans**
I'm going to experiment a bit with some resolvers next.

### CODE REVIEW

File: runtime/new-recipe.js [hunk 1]
```diff
@@ -15,30 +15,106 @@ class Connection extends Edge {
...
+  nameConnection(connection, name) {
+    var idx = this._unnamedConnections.indexOf(connection);
+    assert(idx >= 0);
+    connection._name = name;
+    this._connections[name] = connection;
+    this._unnamedConnections.splice(idx, 1);
+  }
+
+  isResolved() {
+    return this._unnamedConnections.length == 0;
```
Comment (mmandlis) [inline runtime/new-recipe.js]: according to your comment above, resolved Particle should also have an id, so maybe
return this._id !== undefined && ...?

File: runtime/new-recipe.js [hunk 2]
```diff
@@ -15,30 +15,106 @@ class Connection extends Edge {
...
-  get create() {}
-  get connections() {} // ViewConnection*
+  get tags() { return this._tags; } // only tags owned by the view
+  get type() { return this._type; } // nullable
+  get id() { return this._id; }
+  get create() { return this._create; }
+  get connections() { return this._connections } // ViewConnection*
+
+  isResolved() {
+    return this._id !== undefined || this._create == true;
```
Comment (mmandlis) [inline runtime/new-recipe.js]: can a resolved view's type be undefined?

File: runtime/new-recipe.js [hunk 3]
```diff
@@ -15,30 +15,106 @@ class Connection extends Edge {
...
-  get view() {} // View?
-  get particle() {} // never null
+  constructor(name, particle) {
+    assert(particle);
+    this._recipe = particle._recipe;
+    this._name = name;
+    this._tags = [];
+    this._type = undefined;
+    this._direction = undefined;
+    this._particle = particle;
```
Comment (mmandlis) [inline runtime/new-recipe.js]: nit - maybe add
this._view = undefined;
?

File: runtime/new-recipe.js [hunk 4]
```diff
@@ -58,9 +134,36 @@ class Slot extends Node {
...
+  
   get slotConnections() {} // SlotConnection*
-  get viewConnections() {} // ViewConnection*
+  
+  get viewConnections() {
+    var viewConnections = [];
+    this._particles.forEach(particle => {
+      viewConnection.push(...Object.entries(particle.connections));
+      viweConnection.push(...particle._unnamedConnections);
+    });
```
Comment (mmandlis) [inline runtime/new-recipe.js]: nit- missing:
return viewConnections;

File: runtime/recipe.js [hunk 5]
```diff
@@ -114,4 +115,54 @@ class RecipeBuilder {
...
+    this.currentViewConnection = viewConnection;
+    return this;
+  }
+  connectConstraint(name, constraintName) {
+    var view = this.constraints[constraintName];
+    if (!view) {
+      var view = this.recipe.newView();
+      this.constraints[constraintName] = view;
+    }
+    var viewConnection = this.currentParticle.addConnectionName(name);
```
Comment (mmandlis) [inline runtime/recipe.js]: nit - constraintName instead of name?

File: runtime/new-recipe.js [hunk 6]
```diff
@@ -15,30 +15,106 @@ class Connection extends Edge {
...
+  get connections() { return this._connections; } // {parameter -> ViewConnection}
+
+  addUnnamedConnection() {
+    var connection = new ViewConnection(undefined, this);
+    this._unnamedConnections.push(connection);
+    return connection;
+  }
+
+  addConnectionName(name) {
+    this._connections[name] = new ViewConnection(name, this);
```
Comment (mmandlis) [inline runtime/new-recipe.js]: nit - maybe should assert that no connection with this name already exists?

File: runtime/new-recipe.js [hunk 7]
```diff
@@ -15,30 +15,106 @@ class Connection extends Edge {
...
+    if (this._type !== undefined) {
+      if (view._type == undefined)
+        view._type = this._type;
+      else 
+        assert(this_type == view._type);
+    } else if (view._type !== undefined) {
+        this._type = view._type;
+    }
+    this._view = view;
+    this._view.connections.push(this);
```
Comment (mmandlis) [inline runtime/new-recipe.js]: nit: it's this._view._connections

File: runtime/new-recipe.js [hunk 8]
```diff
@@ -15,30 +15,106 @@ class Connection extends Edge {
...
 }
 
 class ViewConnection extends Connection {
-  get name() {} // Parameter name?
-  get tags() {}
-  get type() {}
-  get direction() {} // in/out
-  get view() {} // View?
-  get particle() {} // never null
+  constructor(name, particle) {
```
Comment (mmandlis) [inline runtime/new-recipe.js]: if ViewConnection was to be created directly, it won't be added to the particle's connections map. 
is there a place to assert to verify this doesn't happen? or maybe just leave a comment here.

File: runtime/new-recipe.js (hunk 6 above)
Comment (shans) [inline runtime/new-recipe.js]: Done.

File: runtime/new-recipe.js (hunk 1 above)
Comment (shans) [inline runtime/new-recipe.js]: done

File: runtime/new-recipe.js (hunk 2 above)
Comment (shans) [inline runtime/new-recipe.js]: Don't think so.

File: runtime/new-recipe.js (hunk 8 above)
Comment (shans) [inline runtime/new-recipe.js]: These classes aren't exported directly so they aren't easily constructable. I probably should write some documentation on usage of this module, but I might wait until it's bedded down a bit more.

File: runtime/new-recipe.js (hunk 7 above)
Comment (shans) [inline runtime/new-recipe.js]: the connections getter should work here shouldn't it?

File: runtime/new-recipe.js (hunk 3 above)
Comment (shans) [inline runtime/new-recipe.js]: done

File: runtime/new-recipe.js (hunk 4 above)
Comment (shans) [inline runtime/new-recipe.js]: done

File: runtime/recipe.js (hunk 5 above)
Comment (shans) [inline runtime/recipe.js]: no, the name of the connection is 'name'. The constraintName ties the various connections together.

**Comment by shans** [comment]
@dstockwell take a look at proposed new API?

**Comment by dstockwell** [comment]
> @dstockwell take a look at proposed new API?

👍

File: runtime/new-recipe.js [hunk 9]
```diff
@@ -5,40 +5,175 @@
...
+    this._consumedSlots = [];
+    this._connections = {};
+    this._unnamedConnections = [];
+  }
+
+  clone(recipe, cloneMap) {
+    var particle = new Particle(recipe, this._name);
+    particle._id  = this._id;
+    particle._tags = this._tags.slice();
+    particle._providedSlots = this._providedSlots.map(slot => slot.clone(recipe)); // ?
```
Comment (dstockwell) [inline runtime/new-recipe.js]: // ? = ?

File: runtime/new-recipe.js [hunk 10]
```diff
@@ -5,40 +5,175 @@
...
+    this._tags = [];
+    this._type = undefined;
+    this._direction = undefined;
+    this._particle = particle;
+    this._view = undefined;
+  }
+
+  clone(particle, cloneMap) {
+    var viewConnection = new ViewConnection(this._name, particle);
+    viewConnection._tags = this._tags.slice();
```
Comment (dstockwell) [inline runtime/new-recipe.js]: `[...this._tags]` seems cleaner now

File: runtime/new-recipe.js (hunk 9 above)
Comment (shans) [inline runtime/new-recipe.js]: ?

File: runtime/new-recipe.js (hunk 10 above)
Comment (shans) [inline runtime/new-recipe.js]: done.

**Pull request merged**